src/cryptonote_core/service_node_list.cpp

```diff
@@ -372,6 +365,171 @@ namespace service_nodes
     return money_transferred;
   }
 
+  bool tx_get_staking_components(cryptonote::transaction_prefix const &tx, staking_components *contribution)
+  {
+    staking_components contribution_unused_ = {};
+    if (!contribution) contribution = &contribution_unused_;
+    if (!cryptonote::get_service_node_pubkey_from_tx_extra(tx.extra, contribution->service_node_pubkey))
+      return false; // Is not a contribution TX don't need to check it.
+
+    if (!cryptonote::get_service_node_contributor_from_tx_extra(tx.extra, contribution->address))
+      return false;
+
+    if (!cryptonote::get_tx_secret_key_from_tx_extra(tx.extra, contribution->tx_key))
+    {
+      LOG_PRINT_L1("TX: There was a service node contributor but no secret key in the tx extra");
```

Comment:
Would be useful to include the tx hash here as well.

Reply:
Added